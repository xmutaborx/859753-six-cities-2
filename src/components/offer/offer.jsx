import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import Operations from '../../store/operations';
import {PROP_TYPES_OFFERS_LIST, PROP_TYPES_COMMENTS_LIST, PROP_TYPES_USER_DATA} from '../../constants/prop-types';
import {correctRating} from '../../helpers/helpers';

import Header from '../header/header.jsx';
import FeedbackList from '../feedback-list/feedback-list.jsx';
import CitiesMap from '../cities-map/cities-map.jsx';
import OffersList from '../offers-list/offers-list.jsx';
import FeedbackForm from '../feedback-form/feedback-form.jsx';

class Offer extends React.PureComponent {
  componentDidMount() {
    this.props.getComments(this.props.match.params.id);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.match.params.id !== this.props.match.params.id) {
      window.scrollTo({
        top: 0,
        behavior: `smooth`
      });
      this.props.getComments(this.props.match.params.id);
    }
  }

  render() {
    const offerId = Number(this.props.match.params.id);

    const [currentOffer] = this.props.offers.filter((it) => it.id === offerId);
    const nearOffers = this.props.offers.filter((it) => it.city.name === currentOffer.city.name && it.id !== offerId).slice(0, 3);
    const allOffers = nearOffers.concat([currentOffer]);

    if (!currentOffer) {
      return null;
    }

    return (
      <div className="page">
        <Header />
        <main className="page__main page__main--property">
          <section className="property">
            <div className="property__gallery-container container">
              <div className="property__gallery">
                {currentOffer.images.slice(0, 6).map((it) => (
                  <div className="property__image-wrapper" key={it}>
                    <img className="property__image" src={it} alt="Photo" />
                  </div>
                ))}
              </div>
            </div>
            <div className="property__container container">
              <div className="property__wrapper">
                {currentOffer.is_premium
                && <div className="property__mark">
                  <span>Premium</span>
                </div>}
                <div className="property__name-wrapper">
                  <h1 className="property__name">
                    {currentOffer.title}
                  </h1>
                  <button
                    className={`property__bookmark-button button` + (currentOffer.is_favorite ? ` property__bookmark-button--active` : ``)}
                    type="button"
                    onClick={() => this.props.toggleFavorites(offerId, !currentOffer.is_favorite)}
                  >
                    <svg className="property__bookmark-icon" width={31} height={33}>
                      <use xlinkHref="#icon-bookmark" />
                    </svg>
                    <span className="visually-hidden">To bookmarks</span>
                  </button>
                </div>
                <div className="property__rating rating">
                  <div className="property__stars rating__stars">
                    <span style={{width: `${correctRating(currentOffer.rating)}%`}} />
                    <span className="visually-hidden">{currentOffer.rating}</span>
                  </div>
                  <span className="property__rating-value rating__value">{currentOffer.rating}</span>
                </div>
                <ul className="property__features">
                  <li className="property__feature property__feature--entire">
                    {currentOffer.type}
                  </li>
                  <li className="property__feature property__feature--bedrooms">
                    {currentOffer.bedrooms} Bedrooms
                  </li>
                  <li className="property__feature property__feature--adults">
                    Max {currentOffer.max_adults} {currentOffer.max_adults > 1 ? `adults` : `adult`}
                  </li>
                </ul>
                <div className="property__price">
                  <b className="property__price-value">â‚¬{currentOffer.price}</b>
                  <span className="property__price-text">night</span>
                </div>
                <div className="property__inside">
                  <h2 className="property__inside-title">{`What's inside`}</h2>
                  <ul className="property__inside-list">
                    {currentOffer.goods.map((it) => (
                      <li className="property__inside-item" key={it}>
                        {it}
                      </li>
                    ))}
                  </ul>
                </div>
                <div className="property__host">
                  <h2 className="property__host-title">Meet the host</h2>
                  <div className="property__host-user user">
                    <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                      <img className="property__avatar user__avatar" src={currentOffer.host.avatar_url} width="74" height="74" alt="Host avatar" />
                    </div>
                    <span className="property__user-name">
                      {currentOffer.host.name}
                    </span>
                    {currentOffer.host.is_pro && (
                      <span className="property__user-status">
                        Pro
                      </span>
                    )}
                  </div>
                  <div className="property__description">
                    <p className="property__text">
                      {currentOffer.description}
                    </p>
                  </div>
                </div>
                <section className="property__reviews reviews">
                  <FeedbackList comments={this.props.comments} />
                  {this.props.userData.id && <FeedbackForm id={offerId} />}
                </section>
              </div>
            </div>
            <section className="property__map map">
              <CitiesMap offersList={allOffers} offerId={offerId} />
            </section>
          </section>

          <div className="container">
            <section className="near-places places">
              <h2 className="near-places__title">Other places in the neighbourhood</h2>
              <OffersList offers={nearOffers} nearMode={true} />
            </section>
          </div>

        </main>
      </div>
    );
  }
}

Offer.propTypes = {
  offers: PROP_TYPES_OFFERS_LIST,
  comments: PROP_TYPES_COMMENTS_LIST,
  userData: PropTypes.shape(PROP_TYPES_USER_DATA),
  getComments: PropTypes.func.isRequired,
  toggleFavorites: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    })
  }),
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  offers: state.offers,
  comments: state.comments,
  userData: state.userData,
});

const mapDispatchToProps = (dispatch) => ({
  getComments: (id) => dispatch(Operations.getComments(id)),
  toggleFavorites: (id, status) => dispatch(Operations.toggleFavorites(id, status)),
});

export {Offer};
export default connect(mapStateToProps, mapDispatchToProps)(Offer);
