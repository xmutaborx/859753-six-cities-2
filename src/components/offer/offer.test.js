import React from 'react';
import Enzyme, {shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16/build";
import toJson from "enzyme-to-json";

import {Offer} from './offer.jsx';
import OFFER_MOCK from '../../mocks/offer-mock';
import COMMENT_MOCK from '../../mocks/comment-mock';
import USER_DATA from '../../mocks/user-data-mock';

Enzyme.configure({adapter: new Adapter()});

const mock = {
  params: {
    id: `1`,
  },
};

it(`Offer Page correctly renders`, () => {
  const tree = shallow(<Offer
    offers={OFFER_MOCK}
    comments={COMMENT_MOCK}
    userData={USER_DATA}
    getComments={() => {}}
    toggleFavorites={() => {}}
    match={mock}
  />);

  expect(toJson(tree)).toMatchSnapshot();
});
