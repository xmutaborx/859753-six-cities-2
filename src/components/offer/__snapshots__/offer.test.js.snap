// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offer Page correctly renders 1`] = `
<div
  className="page"
>
  <Connect(Header) />
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
            key="img/1.png"
          >
            <img
              alt="Photo"
              className="property__image"
              src="img/1.png"
            />
          </div>
          <div
            className="property__image-wrapper"
            key="img/2.png"
          >
            <img
              alt="Photo"
              className="property__image"
              src="img/2.png"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Beautiful & luxurious studio at great location
            </h1>
            <button
              className="property__bookmark-button button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="property__bookmark-icon"
                height={33}
                width={31}
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "96%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                4.8
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              4.8
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              apartment
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              3
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              4
               
              adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              â‚¬
              120
            </b>
            <span
              className="property__price-text"
            >
              night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
                key="Heating"
              >
                Heating
              </li>
              <li
                className="property__inside-item"
                key="Kitchen"
              >
                Kitchen
              </li>
              <li
                className="property__inside-item"
                key="Cable TV"
              >
                Cable TV
              </li>
              <li
                className="property__inside-item"
                key="Washing machine"
              >
                Washing machine
              </li>
              <li
                className="property__inside-item"
                key="Coffee machine"
              >
                Coffee machine
              </li>
              <li
                className="property__inside-item"
                key="Dishwasher"
              >
                Dishwasher
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src="img/1.png"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
              <span
                className="property__user-status"
              >
                Pro
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <FeedbackList
              comments={
                Array [
                  Object {
                    "comment": "test comment",
                    "date": "08.12.2019",
                    "id": 69,
                    "rating": 5,
                    "user": Object {
                      "avatar_url": "src",
                      "id": 42,
                      "is_pro": true,
                      "name": "Name",
                    },
                  },
                ]
              }
            />
            <Connect(WithFeedbackForm)
              id={1}
            />
          </section>
        </div>
      </div>
      <section
        className="property__map map"
      >
        <Connect(CitiesMap)
          offerId={1}
          offersList={
            Array [
              Object {
                "bedrooms": 3,
                "city": Object {
                  "location": Object {
                    "latitude": 52.370216,
                    "longitude": 4.895168,
                    "zoom": 10,
                  },
                  "name": "Amsterdam",
                },
                "description": "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.",
                "goods": Array [
                  "Heating",
                  "Kitchen",
                  "Cable TV",
                  "Washing machine",
                  "Coffee machine",
                  "Dishwasher",
                ],
                "host": Object {
                  "avatar_url": "img/1.png",
                  "id": 3,
                  "is_pro": true,
                  "name": "Angelina",
                },
                "id": 2,
                "images": Array [
                  "img/1.png",
                  "img/2.png",
                ],
                "is_favorite": false,
                "is_premium": false,
                "location": Object {
                  "latitude": 52.35514938496378,
                  "longitude": 4.673877537499948,
                  "zoom": 8,
                },
                "max_adults": 2,
                "preview_image": "img/1.png",
                "price": 400,
                "rating": 5,
                "title": "Beautiful & luxurious studio at great location",
                "type": "apartment",
              },
              Object {
                "bedrooms": 3,
                "city": Object {
                  "location": Object {
                    "latitude": 52.370216,
                    "longitude": 4.895168,
                    "zoom": 10,
                  },
                  "name": "Amsterdam",
                },
                "description": "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.",
                "goods": Array [
                  "Heating",
                  "Kitchen",
                  "Cable TV",
                  "Washing machine",
                  "Coffee machine",
                  "Dishwasher",
                ],
                "host": Object {
                  "avatar_url": "img/1.png",
                  "id": 3,
                  "is_pro": true,
                  "name": "Angelina",
                },
                "id": 1,
                "images": Array [
                  "img/1.png",
                  "img/2.png",
                ],
                "is_favorite": false,
                "is_premium": false,
                "location": Object {
                  "latitude": 52.35514938496378,
                  "longitude": 4.673877537499948,
                  "zoom": 8,
                },
                "max_adults": 4,
                "preview_image": "img/1.png",
                "price": 120,
                "rating": 4.8,
                "title": "Beautiful & luxurious studio at great location",
                "type": "apartment",
              },
            ]
          }
        />
      </section>
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <Connect(WithActiveItem)
          nearMode={true}
          offers={
            Array [
              Object {
                "bedrooms": 3,
                "city": Object {
                  "location": Object {
                    "latitude": 52.370216,
                    "longitude": 4.895168,
                    "zoom": 10,
                  },
                  "name": "Amsterdam",
                },
                "description": "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.",
                "goods": Array [
                  "Heating",
                  "Kitchen",
                  "Cable TV",
                  "Washing machine",
                  "Coffee machine",
                  "Dishwasher",
                ],
                "host": Object {
                  "avatar_url": "img/1.png",
                  "id": 3,
                  "is_pro": true,
                  "name": "Angelina",
                },
                "id": 2,
                "images": Array [
                  "img/1.png",
                  "img/2.png",
                ],
                "is_favorite": false,
                "is_premium": false,
                "location": Object {
                  "latitude": 52.35514938496378,
                  "longitude": 4.673877537499948,
                  "zoom": 8,
                },
                "max_adults": 2,
                "preview_image": "img/1.png",
                "price": 400,
                "rating": 5,
                "title": "Beautiful & luxurious studio at great location",
                "type": "apartment",
              },
            ]
          }
        />
      </section>
    </div>
  </main>
</div>
`;
